File: disk_io.pas
{ Low level disk I/O module skeleton for Petit FatFs (C)ChaN, 2014 }

unit disk_io;

{$mode delphi}
interface

const
  { Drive not initialized }
  STA_NOINIT = $01;
  { No medium in the drive }
  STA_NODISK = $02;

type
  DSTATUS = byte;
  DRESULT = (
    { Function succeeded }
    RES_OK = 0,
    { Disk error }
    RES_ERROR,
    { Not ready }
    RES_NOTRDY,
    { Invalid parameter }
    RES_PARERR
    );

{ Initialize Disk Drive }
function disk_initialize(): DSTATUS;

{ Read Partial Sector

  @param(buff Pointer to the destination object)
  @param(sector Sector number (LBA))
  @param(offset Offset in the sector)
  @param(count Byte count (bit15:destination)) }
function disk_readp(buff: Pointer; sector: DWORD; offset: DWORD; Count: DWORD): DRESULT;

{ Write Partial Sector

  @param(buff Pointer to the data to be written,
    NULL:Initiate/Finalize write operation)
  @param(sc Sector number (LBA) or Number of bytes to send) }
function disk_writep(buff: Pointer; sc: DWORD): DRESULT;

implementation

function disk_initialize(): DSTATUS;
var
  stat: DSTATUS;
begin
  (* Put your code here *)
end;

function disk_readp(buff: Pointer; sector: DWORD; offset: DWORD; Count: DWORD): DRESULT;
var
  res: DRESULT;
begin
  (* Put your code here *)

  Result := res;
end;

function disk_writep(buff: Pointer; sc: DWORD): DRESULT;
var
  res: DRESULT;
begin
  if buff = nil then
  begin
    if sc <> 0 then
    begin
      (* Initiate write process *)
    end
    else
    begin
      (* Finalize write process *)
    end;
  end
  else
  begin
    (* Send data to the disk *)
  end;

  Result := res;
end;

end.

File: examples/raspi/c_org/diskio.c
/*-----------------------------------------------------------------------*/
/* Low level disk I/O module skeleton for Petit FatFs (C)ChaN, 2014      */
/*-----------------------------------------------------------------------*/

#include "diskio.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

static int GDH = 0;
extern char *SD_DEV;

/*-----------------------------------------------------------------------*/
/* Initialize Disk Drive                                                 */
/*-----------------------------------------------------------------------*/

DSTATUS disk_initialize (void)
{
	DSTATUS stat;

	// Put your code here
        if (GDH) close(GDH);

	GDH = open(SD_DEV, O_RDWR);
	if (GDH == -1)
	    stat = STA_NOINIT;
	else
	    stat = 0;
	return stat;
}

/*-----------------------------------------------------------------------*/
/* Read Partial Sector                                                   */
/*-----------------------------------------------------------------------*/

DRESULT disk_readp (
	BYTE* buff,		/* Pointer to the destination object */
	DWORD sector,	/* Sector number (LBA) */
	UINT offset,	/* Offset in the sector */
	UINT count		/* Byte count (bit15:destination) */
)
{
	DRESULT res;
	size_t br;

	// Put your code here
	lseek(GDH, sector*512+offset, SEEK_SET);
	br = read(GDH, buff, count);
	if (br == count)
	    res = 0;
	else
	    res = 1;
	

	return res;
}



/*-----------------------------------------------------------------------*/
/* Write Partial Sector                                                  */
/*-----------------------------------------------------------------------*/

DRESULT disk_writep (
	const BYTE* buff,	/* Pointer to the data to be written, NULL:Initiate/Finalize write operation */
	DWORD sc		/* Sector number (LBA) or Number of bytes to send */
)
{
	DRESULT res;


	if (!buff) {
		if (sc) {
			// Initiate write process
			lseek(GDH, 512*sc, SEEK_SET);

		} else {
			// Finalize write process
		}
		res = 0;
	} else {

		// Send data to the disk
		if (write(GDH, buff, sc) == sc)
		    res = 0;
		else
		    res = 1;
                sync();            
	}

	return res;
}

